src/main/scala/diplomacy/Nodes.scala

```diff
@@ -259,27 +733,157 @@ trait OutwardNode[DO, UO, BO <: Data] extends BaseNode
     accPO += ((index, node, binding, p, sourceInfo))
   }
 
-  protected[diplomacy] lazy val oBindings = { oRealized = true; accPO.result() }
+  /** Ends the binding accumulation stage and returns all the output bindings to this node.
+    *
+    * Evaluating this lazy val will mark the outward bindings as frozen,
+    * preventing subsequent bindings from being created via [[oPush]].
+    *
+    * The bindings are each a tuple of:
+    * - numeric index of this binding in the other end of [[InwardNode]].
+    * - [[InwardNode]] on the other end of this binding
+    * - [[NodeBinding]] describing the type of binding
+    * - a [[Parameters]] instance
+    * - [[SourceInfo]] for source-level error reporting
+    */
+  protected[diplomacy] lazy val oBindings: Seq[(Int, InwardNode[DO, UO, BO], NodeBinding, Parameters, SourceInfo)] = { oRealized = true; accPO.result() }
 
+  /** resolved STAR binding of outward nodes: how many connections the star represents. */
   protected[diplomacy] val oStar: Int
+
+  /** a map to convert Node binding index to port range.
+    *
+    * index: the index of connected [[InwardNode]]
+    * element: port range of connected [[InwardNode]]
+    * */
   protected[diplomacy] val oPortMapping: Seq[(Int, Int)]
+
+  /** "Forward" an output connection through this node so that the node can be removed from the graph.
+    *
+    *  @return None if no forwarding is needed.
+    */
   protected[diplomacy] def oForward(x: Int): Option[(Int, OutwardNode[DO, UO, BO])] = None
-  protected[diplomacy] val uoParams: Seq[UO] // from connected nodes
-  protected[diplomacy] val doParams: Seq[DO] // from this node
+  /** Upstream outward parameters.
+    *
+    * Generated from the other nodes that this is connected to.
+    * For specific diagram, see Parameter propagation flow on the top.
+    */
+  protected[diplomacy] val uoParams: Seq[UO]
+  /** Downstream outward parameters.
+    *
+    * Generated from this node.
```

Comment:
Generated by this node and sent downstream to the nodes connected to the outward side of this node.